
matrixplot_modify<-function(data, mapping, pts=list(), smt=list(), ...){
    ggplot(data = data, mapping = mapping, ...) + 
        do.call(geom_point, pts) +
        do.call(geom_smooth, smt) 
}


LFQ_wrapper<-function(maxquant_data,expdesign){
  
  ### Import input MaxQuant proteinGroups.txt file
  maxquant_output<- maxquant_data
  
  ## Optional experimental structure file
  exp_design <- expdesign
  
  ######========= DATA PREPROCESSING =========#######
  # Filter proteins by removing
  # 1. Reverse sequences
  # 2. Potential contaminants
  # 3. Proteins only identified by sites
  # 4. Proteins identified by single peptide
  data <- maxquant_output %>% 
    dplyr:::filter(Reverse!="+", Only.identified.by.site!="+", Razor...unique.peptides>=2)
  
  ## Make data unique by combining  gene names and protein ids in rows where either of them is absent
  data_unique<- DEP:::make_unique(data,"Gene.names","Protein.IDs",delim=";")
  
  #####======= Gnerating Summerised Experiment Object for differential expression analysis =======######
  
  ## get the location of LFQ intensity columns
  lfq_columns<-grep("LFQ.", colnames(data_unique))
  
  ## Convert the protein dataframe into Summerised experiment object for further analysis
  data_se<-DEP:::make_se(data_unique,lfq_columns,exp_design)
  
  
  ## Remove the rows with lot of missing values
  ## This is important step and depends of "thr" option
  ## If thr=0, it means all the replicates in each sample should have valid values, no missing values is allowed
  ## thr=1 means atleast 2 out 3 replicates should have valid values in each sample 
  data_filter<-DEP:::filter_missval(data_se, thr = 1)
  
  # meanSdPlot(data_filter) This can be optional
  
  ######============== Normalisation ===========########
  ## This function uses variance stabilizing transofmation (vsn) function for background correction 
  ## Not much useful for proteomics datasets
  data_norm<-DEP:::normalize_vsn(data_filter)
  #meanSdPlot(data_norm)# Can be optional
  
  ######================ MISSING VALUE IMPUTATION ===============##########
  ## As for this dataset protein intensity plot indicated that proteins with missing value have on average low intensities,
  ## I choose missing value imputation to be one of MNAR method 
  ## Following code will impute missing value with algorithm similar to perseus
  ## i.e. random values drawn from normal distribution of 1.8 SD apart with the width of 0.3
  data_imp_man<-DEP:::impute(data_filter,fun="man",shift=1.8,scale=0.3)
  

  ######============= DIFFERENTIAL EXPRESSION ANALYSIS ============###########
  
  ## This test uses protein-wise linear model with emperical Bayes statistics (used in R package limma)
  ## Limma is modified version of t-test and widely used for transcriptomics analysis
  ## Number of different options available such as "all", "control" and "man" to compare various conditions
  ## "all"- compares all pairwise comparisons
  ## "control"- asks to specify the control sample and compares every other sample against control
  ## "man"- manually specify which condition to test 
  
  #data_diff_all_contrasts<-test_diff(data_imp_man,type='all')
  # If the data has multiple pairwise comparisions we need to generate volcano plot for each comparison
  # "comparison" is a character vectors that stores the name of all comparison and captured from
  #  the message generated by "test_diff" function
  # Example of message "Tested_contrasts: "Control_vs_Condition1", "Control_vs_Condition2""
  comparisons<-capture.output(data_diff_all_contrasts<-test_diff(data_imp_man,type='all'),type = "message")
  
  ## Remove "Tested contrasts:"
  comparisons<-gsub(".*: ","",comparisons)
  ## Split conditions into character vector
  comparisons<-unlist(strsplit(comparisons,","))
  ## Remove leading and trailing spaces
  comparisons<-trimws(comparisons)
  
  ## Mark significantly different proteins
  ## Input needed to define the threshold
  ## "alpha"- adjusted p-value cutoff
  ## "lfc"- log2(Fold Change) cutoff
  alpha<- 0.05
  lfc<- 1
  param<- data.frame(alpha, lfc)
  dep<-DEP:::add_rejections(data_diff_all_contrasts,alpha = 0.05,lfc = log2(1))
  

  ## Plot multiple scatterplots
  ## First get the LFQ expression data
  ## Use ggpairs function from GGally library to generate multiple matrix plots
  paired_data<-as.data.frame(assay(dep))
  

  
  
  ######=============== ENRICHMENT ANALYSIS ============== ##########
  ## Protein Set Enrichment Analysis is based on EnrichR
  ## Need to specify the databases to perform enrichment test
  
  ## Gene Ontology Enrichment
 # gsea_results_GO <- test_gsea(dep)
  
  
  ## KEGG enrichment
 # results_kegg<- test_gsea(dep,databases = c("KEGG_2016"))
  
  
  ####=============== Write Results ===========#######
  data_result<-get_results(dep)
  # write.csv(data_result,"LFQ_results.csv",row.names = FALSE)
  
 save.image (file="data/lfq_results.RData")
return(dep)
  
}

# get_pdf_plot<-function(type_of_plot){
#   pdf(paste0(type_of_plot,".pdf",sep="")
#       plot(type_of_plot)
#       dev.off()
# }

coef_variation<-function(x){
  coef=sd(x)/mean(x)
}

#### Plot CVs

plot_cvs<-function(se) {
  
  ## backtransform data
  untransformed_intensity<- 2^(assay(se))
  exp_design<-colData(se)

### merge untransformed to exp design and calculate cvs
  
  cvs_group<- untransformed_intensity %>% data.frame() %>%
    tibble::rownames_to_column() %>%
    tidyr::gather("ID", "Intensity", -rowname) %>%
    dplyr::left_join(.,data.frame(exp_design), by="ID") %>%
    dplyr::group_by(rowname,condition) %>%
    dplyr::summarise(cvs=coef_variation(Intensity))
  
  ggplot(cvs_group, aes(cvs, color=condition, fill=condition)) +
    geom_histogram(alpha=.5, bins= 20, show.legend = FALSE) +
    facet_wrap(~condition) +
    labs(title= 'Sample Coefficient of Variation', x="Coefficient of Variation", y="Count") +
    theme_DEP2() +
    theme(plot.title = element_text(hjust = 0.5,face = "bold"))   
}


